File: sdk/typescript/src/hash.ts
/**
 * ATB hash-chaining implementation for TypeScript/Node.js.
 *
 * Each event hash is computed as:
 *   SHA256(prevHash + canonicalJSON(event))
 */

import { createHash } from "node:crypto";
import { canonicalize } from "./canonicalize.js";
import { prepareForCanonical } from "./event.js";
import type { Event } from "./event.js";

/** The sentinel previous-hash used for the first event in a bundle. */
export const GENESIS_HASH = "0".repeat(64);

/**
 * Compute the SHA-256 hash for an event given the previous hash.
 *
 * @param event Event to canonicalise and hash.
 * @param prevHash Hex-encoded previous record hash.
 * @returns Hex-encoded SHA-256 hash.
 * @throws TypeError when the event cannot be canonicalised.
 */
export function computeHash(event: Event, prevHash: string): string {
  const canonical = canonicalize(prepareForCanonical(event));
  return createHash("sha256")
    .update(prevHash, "utf8")
    .update(canonical, "utf8")
    .digest("hex");
}

/**
 * Compute and assign hashes for a sequence of events.
 * Mutates each event's `seq` and `prev_hash` fields in-place.
 *
 * @param events Events to chain in-place.
 * @returns Array of hex-encoded SHA-256 hashes, one per event.
 * @throws TypeError when any event cannot be canonicalised.
 */
export function chainEvents(events: Event[]): string[] {
  const hashes: string[] = [];
  let prev = GENESIS_HASH;
  const hasManifest =
    events.length > 0 && events[0].type === "atb.bundle.manifest";
  for (let i = 0; i < events.length; i++) {
    events[i].seq = hasManifest ? Math.max(0, i - 0) : i + 1;
    if (hasManifest && i === 0) {
      events[i].seq = 0;
    } else if (hasManifest) {
      events[i].seq = i;
    }
    events[i].prev_hash = prev;
    const h = computeHash(events[i], prev);
    hashes.push(h);
    prev = h;
  }
  return hashes;
}

File: sdk/python/atb/hash.py
"""
ATB hash-chaining implementation.

Each event hash is computed as:
    SHA256(prev_hash_bytes || canonical_json_bytes)

where prev_hash_bytes is the UTF-8 encoding of the previous event's hex hash,
and canonical_json_bytes is the RFC 8785 canonical JSON of the event object.
"""

from __future__ import annotations

import hashlib
from typing import Any

from atb.canonicalize import canonicalize

#: The sentinel previous-hash used for the first event in a bundle.
GENESIS_HASH = "0" * 64

# Public alias for import convenience.
genesis_hash = GENESIS_HASH


def compute_hash(event: dict[str, Any], prev_hash: str) -> str:
    """Compute the SHA-256 hash for *event* given *prev_hash*.

    Args:
        event: The event dict (must include ``seq``, ``prev_hash``, ``type``, ``data``).
        prev_hash: The hex-encoded hash of the preceding event (or GENESIS_HASH).

    Returns:
        The hex-encoded SHA-256 hash string (64 characters).
    """
    canonical_bytes = canonicalize(event)
    digest = hashlib.sha256(prev_hash.encode("utf-8") + canonical_bytes).hexdigest()
    return digest


def chain_events(events: list[dict[str, Any]]) -> list[str]:
    """Compute and return the hash list for a sequence of events.

    Each event dict is mutated in-place to set ``seq`` and ``prev_hash``.

    Args:
        events: List of event dicts with at least ``type`` and ``data`` keys.

    Returns:
        List of hex-encoded SHA-256 hashes, one per event.
    """
    hashes: list[str] = []
    prev = GENESIS_HASH
    for i, event in enumerate(events):
        event["seq"] = i + 1
        event["prev_hash"] = prev
        h = compute_hash(event, prev)
        hashes.append(h)
        prev = h
    return hashes
